Add command line argument parser

td-extreme opened the pull request.

td-extreme committed:
Add command line argument parser

td-extreme committed:
Core loads server with passed in command line args

The core now loads the server, adds the teapot route and runs.

td-extreme committed:
Put display_welcome back

td-extreme committed:
Added ability to make jar run main

mikeknep commented on src/httpserver/utilities/arguments.clj:
```diff
@@ -0,0 +1,12 @@
+   ["-d" "--directory DIRECTORY"
+    :default  "./"]])
+
+(defn buildMap [args]
```
Use kebab-case instead of camelCase for, well, just about everything in Clojure (except maybe things that interop directly with Java).

Also, this is a pretty vague name for a public function, especially considering almost everything in Clojure is a map.

td-extreme replied:
Wilco!

mikeknep commented on src/httpserver/core.clj:
```diff
@@ -1,6 +1,18 @@
-(ns httpserver.core)
+(ns httpserver.core
+  (:require [httpserver.handlers.teapot :as myteapot]
+            [httpserver.utilities.arguments :as argumenetParser])
```
kebab-case

mikeknep commented on spec/httpserver/utilities/arguments_spec.clj:
```diff
@@ -0,0 +1,24 @@
+           [clojure.tools.cli :refer [parse-opts]]))
+
+(def args '("-p" "5000" "-d" "./site"))
+(def argsEmpty '())
```
I'd like to see you get in the habit of using `let` more often than `def`ing things. You can have nested `let`s and can bind variables with `let` over any number of forms. For example:

``` clojure
(describe "foo"
  (let [args ["-p" "5000" "-d" "./site"]]
    (it "sets port to 5000"
      ...)
    (it "sets directory to ./site"
      (let [something (something-else)]
        ...)))
  (let [empty-args '()]
    ...))
```

td-extreme replied:
okay. Cool!

mikeknep commented on src/httpserver/core.clj:
```diff
@@ -1,6 +1,18 @@
 (def display_welcome "Http Server!")
 
-(defn -main []
-  (println display_welcome))
```
Looks like you can remove the `(def display_welcome "")` line above this as well.

mikeknep commented:
Some minor cleanup around casing and `let`s, but otherwise LGTM. Nice focused PR, and also good job using `require [my.cool.namespace :as namespace'` instead of using `refer` (which is no bueno, IMO)

td-extreme committed:
Cleaned up code

Changed to kebab-case

Removed def in tests and switched to let

removed unused method and test

mikeknep commented on src/httpserver/utilities/arguments.clj:
```diff
@@ -0,0 +1,12 @@
+(ns httpserver.utilities.arguments
+  (:require [clojure.tools.cli :refer [parse-opts]]))
```
Hm, missed this the first time through. I would recommend avoiding `refer` and just aliasing the namespace with `:as` (like you do in your tests). It's functionally the same here, but as a general rule of thumb you want to be consistent about how you're referring in functions from other namespaces and `:as` can't ever bite you, while `refer` could (what if you have a separate function in this namespace named `parse-opts`? Which fn are you calling?)

td-extreme committed:
fixed form order

td-extreme committed:
Chagned :refer to :as
